hswidget/PaletteSelectWidget: refactor

gucio321 opened the pull request:
This PR contains a refactor of palette select (for dc6, dcc and d61 editors) related code.

CHANGELOG:

- [x] fix #299 
  - [x] fix bug fix bug, when palette selector wasn't able to change palette
- [ ] remove unnecessary `rgba` field from dc6 widget state
- [ ] decide, if palette should be saved in as part of editor state
- [ ] rewrite dcc, dc6 and dt1 widget: palette shouldn't be an argument to `Create`. I suppose, it'd be better to add a separated method for setting this argument.

gucio321 committed:
hswidget/PaletteSelectWidget: fix bug, when was unable to close widget with `X` button

gucio321 committed:
all: lintfix

gucio321 committed:
hswidget/PaletteSelectWidget: remove unnecessary closeCB

gucio321 committed:
hswidget/SelectPaletteWidget: rename saveCB -> onSelect

gravestench commented on hswindow/hseditor/hsdc6editor/dc6_editor.go:
```diff
@@ -75,12 +75,10 @@ func (e *DC6Editor) Build() {
 			e.Path.GetUniqueID()+"selectPalette",
 			e.Project,
 			e.config,
+		).IsOpen(&e.selectPalette).OnSelect(
```
I'm not a huge fan of how this was written, but I'll take care of it in another code cleanup PR later.

I'd really like to keep the code from being hard to read :
```golang
if e.selectPaletteWidget == nil {
	id := e.Path.GetUniqueID()+"selectPalette"
	
	onSelectPalette := func(palette *[256]d2interface.Color) {
		e.palette = palette
	}
	
	e.selectPaletteWidget = hswidget.NewSelectPaletteWidget(id, e.Project, e.config).
		IsOpen(&e.selectPalette).
		OnSelect(onSelectPalette)
}
```

gucio321 replied:
see a581bd1

gucio321 committed:
selectpalettewidget: use state to store mpq and project explorers

gucio321 committed:
`hseditor`s: clean Build methods

gravestench commented:
please rebase off of upstream master

gucio321 committed:
all: fix build errors

gucio321 commented:
@gravestench done

gravestench committed:
Merge branch 'master' into issue-299

gravestench reviewed (changes requested):
There's a bug, please fix.

Can not set palette for DC6 (your `issue-299` branch):

https://user-images.githubusercontent.com/936440/124534073-61dcf380-ddc8-11eb-9db0-689a185dd952.mov

On `master`, the feature is working as expected:

https://user-images.githubusercontent.com/936440/124534132-7b7e3b00-ddc8-11eb-8e42-7743a1264b82.mov

gravestench committed:
Merge branch 'master' into issue-299

gravestench committed:
Merge branch 'master' into issue-299

gucio321 commented:
okey, after a small linvestigation, I found that the bug is caused by build method improvement:
```golang
if e.selectPalette {
      // build palette select
}

        dt1Viewer := dt1widget.Create(e.state, e.palette, e.textu
        e.Layout(g.Layout{
                dt1Viewer,
        })
```

so building widget also happens if selectPalette is active
IMO this solution is good and widgets should detect fact of chaning palette
I'll put the `palette` field in widgetState structure, but It'll require a small edits of 3 widgets. I'll work on it soon.

gravestench commented:
I think that your proposed changes, which are related to selecting a palette, would be more coherent inside of this refactor.

gucio321 commented:
yeah, that's what I mean exactly, I'll work on it before this PR will be ready

gucio321 committed:
dc6 and dt1 widgets: fix bug when palette wasn't changed

when palette was changed, the widget didn't detected it. now palette is saved in widgetState and widget detects if palette was changed.

gucio321 committed:
dccwidget: bugfix - palette changer

for more details, see 342fb35

gucio321 committed:
dt1widget: add Palette method

gucio321 committed:
Merge branch 'master' into issue-299

gucio321 committed:
dccwidget: add palette method
